protocol/block-safety-index.md

```diff
@@ -0,0 +1,224 @@
+
+By keeping the last 2 epochs of derived-from mappings in memory,
+the op-node can tell up to which L2 block was derived from a L1 block. And as soon as the L1 finalizes,
+trigger the finality of the L2 blocks that are now irreversibly derived from finalized L1 inputs.
+
+All of this caching logic would become unnecessary,
+if we had a reliable lookup method of this L2-derived-from-L1 mapping data, to act on for any L1 finalization signal.
+
+On top of that: the finality deriver sub-system of the op-node can potentially be completely removed,
+if finality is checked through the op-supervisor API.
+The op-supervisor already knows which L2 block is safe,
+and what L1 block it was derived from, and can thus answer when a L2 block is finalized, given a finalized L1 view.
+
+### The op-proposer consistency problem
+
+The op-proposer has two modes:
+- Wait for finalized L2 block, which we know we can propose, regardless of L1, as it's already final.
+- Wait for a safe L2 block, and make op-node stop-the-world to get consistent derived-from and safe-L2 data.
+
+For safety/stability the finalized-L2 mode is generally used in production.
+But the latter would be faster, and is preferred in testing.
+It can work better if we wouldn't rely on the stop-the-world state, but instead had an index to query,
+to determine which block proposal has to be bound to which L1 blockhash, for reorg safety of L2 proposals.
+
+### The safety indexing problem
+
+Safety indexing is a feature of:
+- explorers: show that a L2 tx became safe at a specific point in L1, possibly even linking the batch data.
+- bridges: determine what L2 blocks to accept for bridging, given some coupling to L1 batch confirmation.
+
+An index, of which L2 block was derived from which L1 block, would help standardize this data,
+and enable every bridge/explorer to interact safely with cross-L2 block data.
+
+# Proposed Solution
+
+Interop, or technically "Isthmus" really, is an upgrade meant to apply to all OP-Stack chains.
+Chains with a dependency-set of 1 (themselves) should not have to run the op-supervisor,
+but should maintain a reliable view of the events and safety of their local chain.
+
+Like the op-supervisor events DB, the history is append-only, except for reorgs.
+This means we could optimize, and store things in an append-only log style system.
+
+The "safedb" in op-node can be replaced with a DB like the above, shared code with the op-supervisor.
+Once the above is done, we can remove the finality mechanism: the op-supervisor signals are used for finalization.
+
+The supervisor is Go, and modular enough that we can embed the single-DB solution into the op-node,
+to serve those less-standard chains with an interop-set of 1.
+This prevents these chains from entering a divergent hardfork path:
+very important to keep the code-base unified and tech-debt low.
+
+The main remaining question is how we bootstrap the data:
```

Comment:
For how long do we actually need to maintain this db? Is there an expiry duration that satisfies all above cases, e.g. the FP challenge window? For most mentioned cases I think a soft rollout could work good enough.